### PR TITLE
support least active request LB and weight  least active reques LB

### PR DESCRIPTION
**Pull request by CodingSinger**
support least active request loadbalancer which always get the host with the least number of active calls.
 * If there is only one host, use the host directly;
 * if there are multiple host with same least active  request value , then choose one based on random.

### CODE REVIEW

File: pkg/upstream/cluster/loadbalancer.go [hunk 1]
```diff
@@ -45,6 +46,7 @@ func init() {
 	}
 	RegisterLBType(types.RoundRobin, rrFactory.newRoundRobinLoadBalancer)
 	RegisterLBType(types.Random, newRandomLoadBalancer)
```
Comment (wangfakang) [inline pkg/upstream/cluster/loadbalancer.go]: it should update [convertLbPolicy]( https://github.com/mosn/mosn/blob/master/pkg/xds/conv/convertxds.go#L985) function  for Istio as a configuration source scenario.

File: pkg/upstream/cluster/loadbalancer.go [hunk 2]
```diff
@@ -129,5 +131,54 @@ func (lb *roundRobinLoadBalancer) HostNum(metadata api.MetadataMatchCriteria) in
...
+		rand:  rand.New(rand.NewSource(time.Now().UnixNano())),
+	}
+}
+
+func (lb *leastActiveLoadBalancer) ChooseHost(context types.LoadBalancerContext) types.Host {
+	healthHosts := lb.hosts.HealthyHosts()
+	if len(healthHosts) == 0 {
+		return nil
+	}
+	// The list of hosts having the same least active reqeuest value
```
Comment (wangfakang) [inline pkg/upstream/cluster/loadbalancer.go]: if single host, we can return early.

Comment (CodingSinger) [reply pkg/upstream/cluster/loadbalancer.go]: good advice.

File: pkg/upstream/cluster/loadbalancer_test.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
+package cluster
```
Comment (taoyuanyuan) [inline pkg/upstream/cluster/loadbalancer_test.go]: could you add a test case of benchmark ？

File: pkg/upstream/cluster/loadbalancer.go [hunk 4]
```diff
@@ -129,5 +131,58 @@ func (lb *roundRobinLoadBalancer) HostNum(metadata api.MetadataMatchCriteria) in
...
+	}
+	// exactly one healthy host, return this host directly
+	if len(healthHosts) == 1 {
+		return healthHosts[0]
+	}
+	// The list of hosts having the same least active request value
+	candicate := make([]types.Host, 0, len(healthHosts))
+	// The least active request value of all hosts
+	leastActive := int64(math.MaxInt64)
+	for _, host := range healthHosts {
```
Comment (taoyuanyuan) [inline pkg/upstream/cluster/loadbalancer.go]: If there are many hosts, it's slow.
We could begin host loop with a random index, and return it directly if the count is zero.

Comment (CodingSinger) [reply pkg/upstream/cluster/loadbalancer.go]: return it directly if the count is zero is a good idea, But is it really necessary to loop hosts with a random index?

File: pkg/upstream/cluster/loadbalancer_test.go (hunk 3 above)
Comment (CodingSinger) [reply pkg/upstream/cluster/loadbalancer_test.go]: done.

File: pkg/upstream/cluster/loadbalancer.go [hunk 5]
```diff
@@ -129,5 +131,58 @@ func (lb *roundRobinLoadBalancer) HostNum(metadata api.MetadataMatchCriteria) in
...
+		} else if active == leastActive {
+			candicate = append(candicate, host)
+		}
+	}
+	//  exactly one host, return this host directly
+	if len(candicate) == 1 {
+		return candicate[0]
+	}
+	// choose one candicate based on the random
+	return candicate[lb.rand.Intn(len(candicate))]
```
[the remark this answers is not in the document]
Comment (CodingSinger) [reply pkg/upstream/cluster/loadbalancer.go]: Thanks for the suggestion.

File: pkg/upstream/cluster/loadbalancer.go (hunk 4 above)
Comment (CodingSinger) [reply pkg/upstream/cluster/loadbalancer.go]: done.

Comment (taoyuanyuan) [reply pkg/upstream/cluster/loadbalancer.go]: such as RoundRobinLB，we first choose host from a random index.
I think that we could do it here.
```
r := random
for i := r; i < len(hosts); i ++ {
}
for i := 0; i < r; i ++ {
}
```

File: pkg/upstream/cluster/loadbalancer.go [hunk 6]
```diff
@@ -129,5 +132,127 @@ func (lb *roundRobinLoadBalancer) HostNum(metadata api.MetadataMatchCriteria) in
...
+	// Check if the original host weights are equal and skip EDF creation if they are
+	if hostWeightsAreEqual(hosts) {
+		return
+	}
+
+	lb.scheduler = newEdfScheduler()
+
+	// Init Edf scheduler with healthy hosts.
+	for _, host := range hosts {
+		lb.scheduler.Add(host, lb.HostWight(host))
```
Comment (JasonRD) [inline pkg/upstream/cluster/loadbalancer.go:241]: Where is the assignment for HostWight?

File: pkg/upstream/cluster/loadbalancer.go [hunk 7]
```diff
@@ -129,5 +132,127 @@ func (lb *roundRobinLoadBalancer) HostNum(metadata api.MetadataMatchCriteria) in
...
+	} else {
+		lb.choice = default_choice
+	}
+	lb.EdfLoadBalancer = newEdfLoadBalancerLoadBalancer(hosts, lb.unweightChooseHost)
+	return lb
+}
+
+func (lb *leastActiveRequestLoadBalancer) unweightChooseHost(context types.LoadBalancerContext) types.Host {
+
+	healthyHosts := lb.hosts.HealthyHosts()
```
Comment (wangfakang) [inline pkg/upstream/cluster/loadbalancer.go:156]: Now `lb.hosts.HealthyHosts()` has been deprecated in this [PR](https://github.com/mosn/mosn/commit/d903bd10ececee4706605f01a73ec193ac9191bd#diff-8d8fb39636bed5042ff869294470b69c), So we should update the logical and you can refer [it](https://github.com/mosn/mosn/blob/master/pkg/upstream/cluster/loadbalancer.go#L72).

Comment (CodingSinger) [reply pkg/upstream/cluster/loadbalancer.go]: done~

File: go.mod [hunk 8]
```diff
@@ -35,7 +35,6 @@ require (
 	golang.org/x/net v0.0.0-20200222125558-5a598a2470a0
 	golang.org/x/sys v0.0.0-20190804053845-51ab0e2deafa
 	golang.org/x/text v0.3.2 // indirect
-	google.golang.org/appengine v1.4.0 // indirect
```
[the remark this answers is not in the document]
Comment (CodingSinger) [reply go.mod]: sorry. my mistake.

File: pkg/upstream/cluster/loadbalancer.go [hunk 9]
```diff
@@ -144,5 +147,133 @@ func (lb *roundRobinLoadBalancer) HostNum(metadata api.MetadataMatchCriteria) in
...
+		lb.mutex.Lock()
+		defer lb.mutex.Unlock()
+		var candicate types.Host
+		// Choose `choice` times and return the best one
+		// See The Power of Two Random Choices: A Survey of Techniques and Results
+		//  http://www.eecs.harvard.edu/~michaelm/postscripts/handbook2001.pdf
+		for cur := 0; cur < int(lb.choice); cur++ {
+
+			randIdx := lb.rand.Intn(healthyHostLen)
+			tempHost := healthyHosts[randIdx]
```
Comment (wangfakang) [inline pkg/upstream/cluster/loadbalancer.go:193]: we should be check the `tempHost.Health() `  health status.

File: pkg/upstream/cluster/loadbalancer.go [hunk 10]
```diff
@@ -144,5 +147,146 @@ func (lb *roundRobinLoadBalancer) HostNum(metadata api.MetadataMatchCriteria) in
...
+	}
+	if total == 1 {
+		// Return directly if there is only one host
+		return targetHosts[0]
+	}
+	for i := 0; i < total; i++ {
+		if lb.scheduler != nil {
+			// do weight selection
+			host := lb.scheduler.Next().(types.Host)
+			lb.scheduler.Add(host, lb.hostWeightFunc(host))
```
Comment (wangfakang) [inline pkg/upstream/cluster/loadbalancer.go:229]: When call `lb.scheduler.Add` function,  the value of  `edf.currentTime `  may already be another item.

File: pkg/upstream/cluster/edf.go [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+// Pick entry with closest deadline and push again
+func (edf *edfSchduler) NextAndPush(weightFunc func(item WeightItem) float64) interface{} {
+	edf.lock.Lock()
+	defer edf.lock.Unlock()
+	if len(edf.items) == 0 {
+		return nil
+	}
+	entry := heap.Pop(&edf.items).(*edfEntry)
+	edf.currentTime = entry.deadline
+	weight := weightFunc(entry.item)
```
Comment (wangfakang) [inline pkg/upstream/cluster/edf.go:55]: Need to deal with abnormal scenarios with a value of 0.

Comment (CodingSinger) [reply pkg/upstream/cluster/edf.go:56]: add some comments for this func, it should be guaranteed by the person who provided the weight func method.

File: pkg/upstream/cluster/loadbalancer.go (hunk 10 above)
Comment (CodingSinger) [reply pkg/upstream/cluster/loadbalancer.go:230]: ok, i have mv this op to the lock block.

File: pkg/upstream/cluster/cluster.go [hunk 12]
```diff
@@ -146,6 +146,8 @@ type clusterInfo struct {
 	lbOriDstInfo         types.LBOriDstInfo
 	tlsMng               types.TLSContextManager
 	connectTimeout       time.Duration
+	lbconfig	         v2.IsCluster_LbConfig
```
Comment (taoyuanyuan) [inline pkg/upstream/cluster/cluster.go:149]: gofmt

File: pkg/upstream/cluster/cluster.go [hunk 13]
```diff
@@ -192,6 +194,10 @@ func (ci *clusterInfo) LbOriDstInfo() types.LBOriDstInfo {
 	return ci.lbOriDstInfo
 }
 
+func (ci *clusterInfo) LbConfig() v2.IsCluster_LbConfig{
```
Comment (taoyuanyuan) [inline pkg/upstream/cluster/cluster.go:197]: go fmt

File: pkg/upstream/cluster/loadbalancer.go [hunk 14]
```diff
@@ -19,6 +19,7 @@ package cluster
 
 import (
 	"math/rand"
+	v2 "mosn.io/mosn/pkg/config/v2"
```
Comment (taoyuanyuan) [inline pkg/upstream/cluster/loadbalancer.go:22]: go imports

File: pkg/upstream/cluster/loadbalancer.go [hunk 15]
```diff
@@ -144,5 +147,146 @@ func (lb *roundRobinLoadBalancer) HostNum(metadata api.MetadataMatchCriteria) in
...
+		return nil
+	}
+	if total == 1 {
+		// Return directly if there is only one host
+		return targetHosts[0]
+	}
+	for i := 0; i < total; i++ {
+		if lb.scheduler != nil {
+			// do weight selection
+			host := lb.scheduler.NextAndPush(lb.hostWeightFunc).(types.Host)
```
Comment (taoyuanyuan) [inline pkg/upstream/cluster/loadbalancer.go:229]: host is unnecessary

File: pkg/upstream/cluster/edf.go [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+func newEdfScheduler(cap int) *edfSchduler {
+	return &edfSchduler{
+		items: make(PriorityQueue, 0, cap),
+	}
+}
+
+// edfEntry is an internal wrapper for item that also stores weight and relative position in the queue.
+type edfEntry struct {
+	deadline float64
+	weight   float64
```
Comment (taoyuanyuan) [inline pkg/upstream/cluster/edf.go:23]: weight not uesed ？

File: pkg/upstream/cluster/loadbalancer.go (hunk 15 above)
Comment (CodingSinger) [reply pkg/upstream/cluster/loadbalancer.go:229]: ok, i will delete it.

File: pkg/upstream/cluster/edf.go (hunk 16 above)
Comment (CodingSinger) [reply pkg/upstream/cluster/edf.go:23]: record the weight of entry.

**Pull request merged**